### DISCUSSION TITLE
Memory limit calculation in policy.xml

### DISCUSSION THREAD

**Discussion by jetpack1116** [ImageMagick/ImageMagick#4226] [Help] [answered]
How to calculate the convert RAM usage for a certain image so that I can adjust the `Memory` limit in `policy.xml` of ImageMagick?
Currently, I used IM ver 7.1.0-1

I want to convert an EPS image to JPG image using parameter `density 600x600`

Example command:
```
magick-density 600x600 "C:\test.eps[0]" -auto-orient -background white -alpha remove -flatten -resize 1920x1920> -depth 8 -quality 80 -quiet +profile "!icc,*" C:\output\test_output.jpg
```

Given:
EPS Image size: 1708 x 1708 
DPI: 72 
Colorspace : sRGB

If this image will be converted with `density 600x600`, the following `width` and `height` pixels is needed:

**width** = (1708 x 600) / 72 = 14233.333 P \~ **14 KP**   
**height** = (1708 x 600) / 72 = 14233.333 P \~ **14 KP**   
Total pixel = 14233.333px \* 14233.333px = 202587777.8 px \~ 203MP   
**RAM** (Color image before jpeg compression) = 203Mp x 24-bit color = **4.9 GB**

Is my calculation correct?

**Comment by snibgo** [comment]
No.

In policy.xml, width and height are the maximum dimensions, in pixels. If your image is 1708x1708 pixels, then those are the dimensions. The density is not relevant.

The memory taken per pixels is more complex. It depends on the IM version, the Q-number, and (for v7) the number of channels per pixel, and whether the build has HDRI. The bit depth of the image in a file is not relevant.

For v7 Q16 non-HDRI, we need 2 bytes per channel per pixel. HDRI needs 4 bytes per channel per pixel. For example, an RGB image has 3 channels, so needs 12 bytes per pixel. If you have 1708*1708 pixels, this is 35 MB.

There is also an overhead per image, and a command may have more than one image in memory at the same time, and decoders and encoders (for reading and writing files) have their own needs.

**Reply by fmw42** [reply]
Density is important if you are trying to process a PDF file. Pixel dimension would be 600dpi * W in inches x 600dpi * H in inches.

**Comment by snibgo** [comment, marked as answer]
Thanks for the correction, fmw42.

As the input is EPS, the density _is_ important, because that determines how many pixels IM has to record. If the rasterized width is 14233 pixels, then use that number in your calculations.

You can insert "+write info:" in the command, immediately after reading the EPS file, to find the number of pixels.

**Reply by jetpack1116** [reply]
I actually follow the calculation of @fmw42  in older post, see link below.

https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=31552#:~:text=Re%3A%20Can%27t%20convert,policy.xml%20file.

I just wonder why @fmw42 multiply the number of pixels by 24?...he just wrote it as 24-bits color, then he got the result in GB...

**Reply by jetpack1116** [reply]
I actually use this ImageMagick-7.1.0-portable-Q16-x86.zip
It says 
```
Portable Win32 static at 16 bits-per-pixel component. Just copy to your host and run (no installer, no Windows registry entries).
```
I am not quite sure where to get that info regarding bytes per channel and how many channel per pixel.
I also assume that it is non HDRI since there is no`HDRI` written over it.

**Reply by jetpack1116** [reply]
running `identify` command to the eps file, I got this info
```
C:¥identify C:\test0001.eps
C:\test0001,eps EPS 1708x1708 1708x1708+0+0 16-bit sRGB 21526B 0.000u 0:00.000
```

**Reply by fmw42** [reply]
That is the pixel dimensions for the default density of 72 dpi.  If you want 600 dpi then multiply 1708\*600/72=14233. So your image would be 14233x14233=202578289 total pixels. But each pixel is 24bits (if color and no transparency). So multiply 202578289\*24=4861878936 so about 5 GB, if I have not made a mistake.

**Reply by jetpack1116** [reply]
@fmw42 
Im just confused with the multiplication of 24bits.
Then the result is in GB (Gigabytes).

But its 24 bits not 24 bytes right?
If it is 24 bits, then shouldn't we divide it by 8 to get the bytes unit?

202578289*24bits=4861878936bits = 607734867bytes = 0.6GB

**Reply by fmw42** [reply]
My mistake. Yes, you are correct. It is 24 bits or 4 bytes. So 202578289\*4 GB.

**Reply by jetpack1116** [reply]
@fmw42 

I see. Thank you for the clarification. 👍 

Actually Im solving this problem wherein EPS to JPG conversion using `-density 300x300` successfully converts.
But if I use `-density 600x600`, conversion fails.
This is the case when I use IMver7.0.8-6.
If I run `identify -list resource`, I got this:

```
C:\imagemagick>identify -list resource
Resource limits:
  Width: 8000P
  Height: 8000P
  List length: -1
  Area: 8.50302GP
  Memory: 3.95953GiB
  Map: 7.91905GiB
  Disk: unlimited
  File: 1536
  Thread: 1
  Throttle: 0
  Time: unlimited
```

Clearly its not the `Memory` that is problematic.
It is the `Width` and `Height` limit problem.

Because if 300x300 ---> Width= 7000P, Height= 7000P  ---> Well within limit, so ok for conversion
If 600x600 -----> Width=14000P, Height=14000P ---> Over the limit, conversion fails.

Now if I use latest IM version, say ver 7.1.0-0...if I `identify -list resource` it, I get:
```
C:\imagemagick_v7.1.0-1>identify -list resource
Resource limits:
  Width: 922.337PP
  Height: 922.337PP
  Area: 17.006GP
  List length: unlimited
  Memory: 7.91906GiB
  Map: 15.8381GiB
  Disk: unlimited
  File: 1536
  Thread: 1
  Throttle: 0
  Time: unlimited
```
Here, Width and Height limit is 922.337PP  (I dont know what PP means...is it PetaPixels?)
So there's no problem in conversion wether I use 300x300 or 600x600.

Thank you for the explanation @fmw42 👍  and @snibgo 👍

**Reply by fmw42** [reply]
Yes, I believe that is correct. So change your policy to increase the width and height limits.

**Reply by snibgo** [reply]
Yes, "PP" is peta pixels. Peta is 1000^5, which is a thousand million million pixels.

But the Area (width * height) limit is 17.006GP, about 17 thousand million pixels, which is a tighter (smaller) limit.
